Snowflake beta (#3722)
braze-inc/braze-docs: _docs/_hidden/private_betas/snowflake_integration.md

{% endalert %}

## What is Braze Cloud Ingestion?

Braze Cloud Ingestion allows you to set up a direct connection from your Snowflake instance to Braze to sync relevant user data. Once synced to Braze, these attributes can be used for personalization or segmentation.

### How it works

With Braze Cloud Ingestion, you set up an integration between your Snowflake instance and Braze app group to sync data on a recurring basis. This sync runs on a schedule you set, and each integration can have a different schedule. 

When a sync runs, Braze will directly connect to your Snowflake and pull all new data from the table, and update the corresponding user profiles on your Braze dashboard. Each time the sync runs, any updated data will be reflected on the user profiles.

### What gets synced

Each time a sync runs, Braze looks for rows that have not previously been synced. We check this using the `UPDATED_AT` column in your table or view. Any rows where `UPDATED_AT` is later than the last synced row will be selected and pulled into Braze.

### Data point usage 

> Each attribute sent for a user will consume one data point. It’s up to the customer to only send the required data. Data point tracking for Cloud Ingestion is equivalent to tracking through the [`/users/track`]({{site.baseurl}}/api/endpoints/user_data/post_user_track#user-track) endpoint. Refer to [Data points]({{site.baseurl}}/user_guide/onboarding_with_braze/data_points/), for more information.

Reviewer: ```suggestion
Each attribute sent for a user will consume one data point. It’s up to you to only send the required data. Data point tracking for Cloud Ingestion is equivalent to tracking through the [`/users/track`]({{site.baseurl}}/api/endpoints/user_data/post_user_track#user-track) endpoint. Refer to [Data points]({{site.baseurl}}/user_guide/onboarding_with_braze/data_points/) for more information.
```

## Data setup recommendations

### Only write new or updated attributes to minimize consumption

We will sync all attributes in a given row, regardless of whether they are the same as what’s currently on the user profile. Given that, we recommend that you only sync attributes that you want to add or update.

### Removing an attribute

> If you want to completely remove an attribute from a user’s profile, you can set it to `null`. If you want an attribute to remain unchanged, you can just not send it to Braze until it’s been updated.

Reviewer: Suggested for clarity:
```suggestion
If you want to completely remove an attribute from a user’s profile, you can set it to `null`. If you want an attribute to remain unchanged, don't send it to Braze until it’s been updated.
```

### Create JSON string from another table

> If you prefer to store each attribute in its own column internally, you need to convert those columns to a JSON string to populate the sync with Braze. To do that,  you can use a query like: 

Reviewer: Extra space here
```suggestion
If you prefer to store each attribute in its own column internally, you need to convert those columns to a JSON string to populate the sync with Braze. To do that, you can use a query like: 
```

```json
CREATE TABLE "PURCHASE_DATA" 
    (purchase_date datetime,
     purchase_amount number,
     quantity number,
     address string);

SELECT TO_JSON(OBJECT_CONSTRUCT (*)) FROM "PURCHASE_DATA";
```

## Product setup

During the beta phase, onboarding to Cloud Ingestion requires some manual integration steps managed by Braze. Follow these steps to set up the integration: 
1. In your Snowflake instance, set up the table(s) or view(s) you want to sync to Braze.
2. Submit a form to Braze with the relevant integration information.
3. Braze will send you a public key to [append to the Snowflake user](https://docs.snowflake.com/en/user-guide/key-pair-auth.html) for authentication.

> 4. Sync starts.

Reviewer: Small nit on step 4 here, but this isn't actually a step that the user takes (unless they need to start the sync, in which case we should reword that). I'd suggest removing it from the list and just make it an ending sentence.

### Set up tables or views

#### Step 1: Set up the table

```json
CREATE DATABASE BRAZE_CLOUD_PRODUCTION;
CREATE SCHEMA BRAZE_CLOUD_PRODUCTION.INGESTION;
CREATE TABLE BRAZE_CLOUD_PRODUCTION.INGESTION.USERS_ATTRIBUTES_SYNC (
     UPDATED_AT datetime not null, 
     EXTERNAL_ID string not null,
     PAYLOAD string not null)
;
```

> You can name the database, schema, and table as you’d like, but the column names should match the above definition.

Reviewer: ```suggestion
You can name the database, schema, and table as you’d like, but the column names should match the preceding definition.
```

```json
CREATE USER BRAZE_INGESTION_USER;

GRANT ROLE BRAZE_INGESTION_ROLE TO USER BRAZE_INGESTION_USER;
```

After this step, you will share connection information with Braze and receive a public key to append to the user. 

#### Step 5: Allow Braze IPs in Snowflake network policy (optional)

Allow the following IP addresses in your Snowflake network policy. For more information on enabling this, see the relevant Snowflake documentation on [modifying a network policy](https://docs.snowflake.com/en/user-guide/network-policies.html#modifying-network-policies). 

| Braze IPs |
| -------- |
| 23.21.118.191 |
| 34.206.23.173 |
| 50.16.249.9 |
| 52.4.160.214 |
| 54.87.8.34 |
| 54.156.35.251 |
| 52.54.89.238 |
| 18.205.178.15 |

### Share connection information with Braze 

Set up your integration by sharing the connection information with Braze through [this form](https://docs.google.com/forms/d/1df0dbw4G_XBkSD_ikT0-zJGg5K6OTOKKe-ARUx9j21M/edit). When you fill out the form, we will ask for the following information:

- Snowflake connection information
  - Account name 
  - Warehouse 
  - Database
  - Table 
  - User 
  - Role
- Braze integration information

>   - App group ID - The ID of the app group where the data should be synced (you can find this ID in the URL if you go to your app group settings page). If you need help, please ask

Reviewer: ```suggestion
  - App group ID - The ID of the app group where the data should be synced (you can find this ID in the URL if you go to your app group settings page). If you need help, please ask.
```

>   - API key - Submit an existing API key to use, or we can create a new one for you

Reviewer: ```suggestion
  - API key - Submit an existing API key to use, or we can create a new one for you.
```

- Sync information

>   - Integration name - A label you can use to identify the integration(s) you set up with your Snowflake instance

Reviewer: ```suggestion
  - Integration name - A label you can use to identify the integration(s) you set up with your Snowflake instance.
```
